openraft/src/testing/suite.rs

```diff
@@ -0,0 +1,1511 @@
+
+        let res = store.apply_to_state_machine(Vec::<&Entry<_>>::new().as_slice()).await;
+
+        let e = res.unwrap_err().into_defensive().unwrap();
+        assert_eq!(ErrorSubject::Logs, e.subject);
+        assert_eq!(Violation::LogsEmpty, e.violation);
+
+        Ok(())
+    }
+
+    pub async fn df_apply_index_eq_last_applied_plus_one(builder: &B) -> anyhow::Result<()> {
+        let store = builder.build().await;
+
+        let entry = blank(3, 1);
+
+        store.apply_to_state_machine(&[&blank(0, 0), &entry]).await?;
+
+        tracing::info!("--- re-apply 1th");
+        {
+            let res = store.apply_to_state_machine(&[&entry]).await;
+
+            let e = res.unwrap_err().into_defensive().unwrap();
+            assert_eq!(ErrorSubject::Apply(LogId { term: 3, index: 1 }), e.subject);
+            assert_eq!(
+                Violation::ApplyNonConsecutive {
+                    prev: Some(LogId { term: 3, index: 1 }),
+                    next: LogId { term: 3, index: 1 },
+                },
+                e.violation
+            );
+        }
+
+        tracing::info!("--- apply 3rd when there is only 1st");
+        {
+            let entry = blank(3, 3);
+            let res = store.apply_to_state_machine(&[&entry]).await;
+
+            let e = res.unwrap_err().into_defensive().unwrap();
+            assert_eq!(ErrorSubject::Apply(LogId { term: 3, index: 3 }), e.subject);
+            assert_eq!(
+                Violation::ApplyNonConsecutive {
+                    prev: Some(LogId { term: 3, index: 1 }),
+                    next: LogId { term: 3, index: 3 },
+                },
+                e.violation
+            );
+        }
+
+        Ok(())
+    }
+
+    pub async fn df_apply_gt_last_applied_id(builder: &B) -> anyhow::Result<()> {
+        let store = builder.build().await;
+
+        let entry = blank(3, 1);
+
+        store.apply_to_state_machine(&[&blank(0, 0), &entry]).await?;
+
+        tracing::info!("--- next apply with last_index+1 but lower term");
+        {
+            let entry = blank(2, 2);
+            let res = store.apply_to_state_machine(&[&entry]).await;
+            assert!(res.is_err());
+
+            let e = res.unwrap_err().into_defensive().unwrap();
+            assert_eq!(ErrorSubject::Apply(LogId { term: 2, index: 2 }), e.subject);
+            assert_eq!(
+                Violation::ApplyNonConsecutive {
+                    prev: Some(LogId { term: 3, index: 1 }),
+                    next: LogId { term: 2, index: 2 },
+                },
+                e.violation
+            );
+        }
+
+        Ok(())
+    }
+}
+
+/// Create a blank log entry for test
+fn blank<D: AppData>(term: u64, index: u64) -> Entry<D> {
+    Entry {
+        log_id: LogId::new(term, index),
+        payload: EntryPayload::Blank,
+    }
+}
+
+/// Block until a future is finished.
+/// The future will be running in a clean tokio runtime, to prevent an unfinished task affecting the test.
```

Comment:
I wonder if this is needed, looks like the current log tests can be executed within a single runtime sequentially

Reply:
Or maybe we can run them concurrently, every test has its own store and state machine, so they should not affect each other, though I don't think this is needed.

Reply:
No it does not looks necessary anymore 🤔

Reply:
Get it, then I will proceed to make the log tests runtime-agnostic by turning `test_all()` and `test_store()` asynchronous:)